Demote local data member

tommyk-gears opened the pull request:
<!--
Contributing to Hazelcast and looking for a challenge? Why don't you check out our open positions?

https://hazelcast.bamboohr.com/jobs
-->

Introduces `com.hazelcast.cluster.Cluster#demoteLocalDataMember` method which transitions the local member from a data/normal member to a lite member. The most dramatic effect might be that a cluster can now transition from a state where it holds data to a state where it does not hold any data, but the cluster is still "alive" (i.e. when demoting the last data member).

The implementation is mostly a blend of member promotion and member shutdown.

Checklist:
- [ ] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [ ] Label `Add to Release Notes` or `Not Release Notes content` set
- [ ] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc

devOpsHazelcast commented:
[![CLA assistant check](https://cla.hazelcast.org:443/pull/badge/not_signed)](https://cla.hazelcast.org:443/hazelcast/hazelcast?pullRequest=24617) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.hazelcast.org:443/hazelcast/hazelcast?pullRequest=24617) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.hazelcast.org:443/check/hazelcast/hazelcast?pullRequest=24617) it.</sub>

devOpsHazelcast commented:
Can one of the admins verify this patch?

devOpsHazelcast commented:
Can one of the admins verify this patch?

vbekiaris commented:
run-lab-run

vbekiaris commented on hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/DemoteDataMemberTest.java:
```diff
@@ -0,0 +1,516 @@
+    public ExpectedException exception = ExpectedException.none();
+
+    @Test
+    public void dataMaster_demoted() {
```
hm.. I have a few reservations about this use case - draining a cluster of all data members after initial partition arrangement is done. I would need to look a bit deeper into this.

tommyk-gears replied:
I understand/share your concerns. Actually, in _my_ use case there is no need to demote the _last_ data member, so from my point of view I'd be happy to just dis-allow it.
I'll leave it for you to decide which option is best from a product perspective :)

vbekiaris reviewed (commented):
@tommyk-gears thanks a lot for the contribution. I am not sure we should allow removing all data members from the cluster after initial partition arrangement is done, but I didn't want to delay posting my comments further, so here they are. Let me know your thoughts around that.

thanks once more for the contribution!

tommyk-gears commented on hazelcast/src/main/java/com/hazelcast/internal/partition/impl/PartitionStateGeneratorImpl.java:
```diff
@@ -59,12 +59,13 @@ public PartitionReplica[][] arrange(Collection<MemberGroup> memberGroups, Intern
+
         if (groups.isEmpty()) {
-            return null;
+            return state;
```
The intention with this change is to allow the last data member in the cluster to be demoted, which requires us to actively migrate from the current state to an empty partition table afaiu. I.e. my intention was that after the demotion completed, the demoted node is _guaranteed to not own any partitions_.

As of now we mainly end up in this case at _shutdown_ of the last data member, and in that case the partition table is left in a slightly "invalid" state and node is allowed to terminate without any migration done (there is nowhere to migrate the data anyway, so it makes sense).

In the case where the cluster contains lite member(s) and 1 data member, and the data member shuts down - I suppose the terminated member is eventually removed from the partition table by the repair task?
The normal case is that there are no lite member, and then the cluster itself is terminated when the last data member is shut down, so the exact state of the partition table is not important, right?

But, if we should not support demotion of the last member we can/should revert this change.

ramizdundar commented on hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/DemoteDataMemberTest.java:
```diff
@@ -0,0 +1,516 @@
+    }
+
+    @Test
+    public void demotion_shouldFail_whenMasterLeaves_duringDemotion() throws Exception {
```
I tried to cancel these comments but somehow GitHub commented them. Please ignore them as they are fixed with your latest commits.

ramizdundar commented on hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/DemoteDataMemberTest.java:
```diff
@@ -0,0 +1,516 @@
+        InternalCompletableFuture<MembersView> future =
+                getOperationService(hz2).invokeOnTarget(ClusterServiceImpl.SERVICE_NAME, op, getAddress(hz3));
+        exception.expect(CompletionException.class);
+        exception.expect(new RootCauseMatcher(IllegalStateException.class));
```
I tried to cancel these comments but somehow GitHub commented them. Please ignore them as they are fixed with your latest commits.

ramizdundar commented on hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/DemoteDataMemberOp.java:
```diff
@@ -0,0 +1,63 @@
+ *
+ * @since 5.4
+ */
+public class DemoteDataMemberOp extends AbstractClusterOperation {
```
IIUC, there is the following ordering guarantee between these 3 operations:
```
DemoteRequestOperation (master) -> DemoteResponseOperation (member) -> DemoteDataMemberOp (master)
```
 Can you add this to docs of the mentioned 3 ops? Especially `DemoteResponseOperation -> DemoteDataMemberOp`  ordering isn't clear from the docs.

tommyk-gears replied:
Addressed in this commit: https://github.com/hazelcast/hazelcast/pull/24617/commits/43670f7ecafcb2ce070608bafd67b0f17eefc06f

tommyk-gears commented on hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/DemoteDataMemberTest.java:
```diff
@@ -0,0 +1,516 @@
+    }
+
+    @Test
+    public void demotion_shouldFail_whenMasterIsSuspected_duringDemotion() throws Exception {
```
After I changed to using `#invokeOnMaster` in `ClusterServiceImpl#demoteLocalDataMember` I lost track of where the returned `MemberView` came from. We need to know that in order to make the validations in `ClusterServiceImpl#updateMembers`.

In https://github.com/hazelcast/hazelcast/pull/24617/commits/42a2a47db43c0dd9f6c1d0c993615a5412ab1d24 I solved this by changing response type of DemoteDataMemberOp to the new type `MembersViewResponse` which also includes the info about which member the `MemberView` comes from along with `MemberView` itself.

This _works_ but it makes the code a bit cluttered - e.g. promotion and demotion are slightly different in this aspect for no apparent reason.

Other possible solutions might be;
* Add such info inside `MemberView`. 
* Also change response type of `PromoteLiteMemberOp` to  `MembersViewResponse` and use `invokeOnMaster` (it would probably be challenging from a compatibility perspective)
* Just skip the validation in `ClusterServiceImpl#updateMembers` after we used invokeOnMaster?

What is your thoughts on this?

tommyk-gears commented on hazelcast/src/main/java/com/hazelcast/internal/partition/operation/DemoteResponseOperation.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.UUID;
+
+/**
+ * Signal that the given member no longer owns any replicas (and its state may be changed to lite member).
```
Addressed in https://github.com/hazelcast/hazelcast/pull/24617/commits/4ce2adcf5663da737128735ff5a88d59e849082a

tommyk-gears commented on hazelcast/src/test/java/com/hazelcast/internal/cluster/impl/DemoteDataMemberTest.java:
```diff
@@ -0,0 +1,516 @@
+    }
+
+    @Test
+    public void dataMember_shouldLooseDataWhenLastDataMember() {
```
Addressed in this commit: https://github.com/hazelcast/hazelcast/pull/24617/commits/dd6c364b93c1e4c2047b3fdbaf7caa4d786151c5

srknzl commented:
run-lab-run

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/20369/Hazelcast-pr-builder-20369.zip), [artifacts](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/20369/Hazelcast-pr-builder-20369.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembersViewResponse.java:1: Line does not match expected header line of '/*'. [Header]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java:614: Line is longer than 130 characters (found 137). [LineLength]
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.3.0:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 2 errors reported by Checkstyle 9.3 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
[ERROR] 
--------------------------
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
--------------------------
[ERROR] 
--------------------------
[ERROR] For more information about the errors and possible solutions, please read the following articles:
--------------------------
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
--------------------------
[ERROR] 
--------------------------
[ERROR] After correcting the problems, you can resume the build with the command
--------------------------
[ERROR]   mvn <args> -rf :hazelcast
--------------------------
</pre>
</details>

vbekiaris commented:
run-lab-run

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/20775/Hazelcast-pr-builder-20775.zip), [artifacts](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/20775/Hazelcast-pr-builder-20775.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
---------ERRORS-----------
--------------------------
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 5.899 s <<< FAILURE! -- in com.hazelcast.HazelcastCompletableFutureAsyncUsageTest
--------------------------
[ERROR] com.hazelcast.HazelcastCompletableFutureAsyncUsageTest.noClassUsesCompletableFuture -- Time elapsed: 5.898 s <<< FAILURE!
--------------------------
[ERROR] Failures: 
--------------------------
[ERROR]   HazelcastCompletableFutureAsyncUsageTest.noClassUsesCompletableFuture:40 Architecture Violation [Priority: MEDIUM] - Rule 'classes should use only CompletableFuture async methods with explicit executor service' was violated (1 times):
--------------------------
[ERROR] Tests run: 4574, Failures: 1, Errors: 0, Skipped: 17
--------------------------
[ERROR] There are test failures.
--------------------------
[ERROR] Tests run: 17, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 192.6 s <<< FAILURE! -- in com.hazelcast.jet.kinesis.KinesisIntegrationTest
--------------------------
[ERROR] com.hazelcast.jet.kinesis.KinesisIntegrationTest.dynamicStream_1Shard_splitsDuringData -- Time elapsed: 126.5 s <<< FAILURE!
--------------------------
[ERROR] Failures: 
--------------------------
[ERROR]   KinesisIntegrationTest.dynamicStream_1Shard_splitsDuringData:324->dynamicStream_splitsDuringData:354->assertMessages:637->AbstractKinesisTest.assertMessages:163->HazelcastTestSupport.assertTrueEventually:1269->HazelcastTestSupport.assertTrueEventually:1165->AbstractKinesisTest.lambda$assertMessages$3:180 Messages for key 243 differ!
--------------------------
[ERROR] Tests run: 26, Failures: 1, Errors: 0, Skipped: 3
--------------------------
[ERROR] There are test failures.
--------------------------
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   HazelcastCompletableFutureAsyncUsageTest.noClassUsesCompletableFuture:40 Architecture Violation [Priority: MEDIUM] - Rule 'classes should use only CompletableFuture async methods with explicit executor service' was violated (1 times):
com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl.checkClusterPartitionRuntimeStates():624 calls CompletableFuture.thenCompose
[INFO] 
[ERROR] Tests run: 4574, Failures: 1, Errors: 0, Skipped: 17
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

JamesHazelcast commented on hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java:
```diff
@@ -630,6 +635,26 @@ void resetMigrationInterceptor() {
+            if (demoteRequestedMembers.add(member)) {
+                logger.info("Demote request of " + member + " is handled");
+            }
+            sendDemoteResponseOperation(member);
```
Is it intentional for this call to be outside the `demoteRequestedMembers.add(member)` conditional check? The operation seems idempotent to me, but I wanted to flag so we can make sure!

tommyk-gears replied:
It is intentional to solve cases where the first `DemoteResponseOperation` is never received on the other end for some reason and a new `DemoteRequestOperation` is sent as a retry.

I updated `DemoteResponseOperation`/ `com.hazelcast.internal.partition.impl.InternalPartitionServiceImpl#onDemoteResponse` so that it is also idempotent.

JamesHazelcast commented on hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java:
```diff
@@ -1942,6 +1977,65 @@ public void run() {
+                final int demoteRequestCount = demoteRequestedMembers.size();
+                if (demoteRequestCount > 0) {
+                    boolean present = false;
+                    List<Member> demotedMembers = new ArrayList<>();
```
Nit-pick: We can initialize the `ArrayList` with the size of `demoteRequestCount`

tommyk-gears replied:
Addressed in https://github.com/hazelcast/hazelcast/pull/24617/commits/c50accbff226306616ef05fbbeba58081b91443d

JamesHazelcast reviewed (commented):
Thanks for your PR @tommyk-gears, it's very well presented! Just a few minor comments from me, nothing major, although I do share Vassilis' concerns surrounding the demotion of the last member of a cluster to Lite ([as discussed here](https://github.com/hazelcast/hazelcast/pull/24617/files#r1231989722)).

Once we've reached some clarity on that (pending Vassilis' decision), I think we'll basically be ready for approval 👍

tommyk-gears commented on hazelcast/src/main/java/com/hazelcast/internal/partition/operation/DemoteRequestOperation.java:
```diff
@@ -0,0 +1,116 @@
+ * @since 5.4
+ */
+public class DemoteRequestOperation
+        extends AbstractPartitionOperation implements MigrationCycleOperation, Versioned {
```
Addressed in https://github.com/hazelcast/hazelcast/pull/24617/commits/c50accbff226306616ef05fbbeba58081b91443d

tommyk-gears commented on hazelcast/src/main/java/com/hazelcast/internal/partition/operation/DemoteResponseOperation.java:
```diff
@@ -0,0 +1,106 @@
+ * @since 5.4
+ */
+public class DemoteResponseOperation
+        extends AbstractPartitionOperation implements MigrationCycleOperation, Versioned {
```
Addressed in https://github.com/hazelcast/hazelcast/pull/24617/commits/c50accbff226306616ef05fbbeba58081b91443d

tommyk-gears commented on hazelcast/src/main/java/com/hazelcast/spi/properties/ClusterProperty.java:
```diff
@@ -699,6 +699,12 @@ private int getWhenNoSSLDetected() {
+     */
+    public static final HazelcastProperty DEMOTE_MAX_WAIT
+            = new HazelcastProperty("hazelcast.member.demote.max.wait", 600, SECONDS);
+
```
Addressed in https://github.com/hazelcast/hazelcast/pull/24617/commits/c50accbff226306616ef05fbbeba58081b91443d

tommyk-gears commented:
> Hi @tommyk-gears, I've discussed with Vassilis and we both agree that a Hazelcast cluster without any data members does not make much sense, so we think it's best if we throw an `IllegalStateException` when an attempt is made to demote the last data member of a cluster. Let us know what you think 👍
> 
> This PR is currently [failing one of our archunit tests](https://github.com/hazelcast/hazelcast/pull/24617#issuecomment-1709944894) that enforces an executor to be defined explicitly for non-async method calls in `CompletableFuture` - so switching `thenCompose(x ->x)` to `thenComposeAsync(x -> x, internalAsyncExecutor)` should make it happy 😄

Awesome, thanks for good feedback! I changed the implementation to throw `IllegalStateException` when trying to demote the last data member, and also reverted the code changes that were made to support demotion of last member.
Also fixed that `thenCompose`-issue.

JamesHazelcast commented:
run-lab-run

JamesHazelcast commented:
run-lab-run

JamesHazelcast commented:
Hi @hazelcast/apis, could someone please review? TIA!

vbekiaris reviewed (approved):
thanks for contributing this enhancement @tommyk-gears!

vbekiaris commented:
thank you @tommyk-gears for your contribution 🏅
